Add `tag-changelog-link` feature

HardikModha opened the pull request:
Should fix #966.

I tried to use the API but I couldn't find a way to fetch details of next and previous tags and fetching all tags at once wasn't feasible. (See discussion #966)

I've taken the following approach. 

1. First, I've extracted all tags, commit ids of those tags.

2. Now, iterating over each tag, I see if the next tag has the different commit id. (Given the fact that we can have multiple tags on the same commit id).  

3. If the next tag has different commit id, then I further check whether the tags are namespaced or not. If the namespace is the same then pick that tag for comparison.

4. If the next tag has same commit id then that tag is skipped and we check remaining tags.

5. For the last tag, I fetch the next page using `fetchDom` and perform the same steps as mentioned above. If a tag to compare is not found on the next page then we skip the tag. 

**Tested it on the following (and many other) repos.**

1. https://github.com/parcel-bundler/parcel/releases (It has multiple, namespaced and normal semver style tags)
2. https://github.com/tensorflow/tensorflow/releases (With Alpha, Beta kind of tags)
3. https://github.com/facebook/react/releases

**The feature works on the following pages:**

1. releases listing: e.g. https://github.com/parcel-bundler/parcel/releases
2. tags listing: e.g. https://github.com/parcel-bundler/parcel/tags
3. Single tag pages: https://github.com/parcel-bundler/parcel/releases/tag/v1.10.3
 
**Limitations:**

1. ~Namespaced tags are only checked within the single page to find it's previous tag~. **Update**: Now we check on the next page, too. 

2. ~For the last tag on the page, we only check on the next page to find the previous tag.~ **Update**: For all the tags on any page, we also check the next page for finding its previous tag. 

3. The tags are not semantically compared. But we can easily implement it.

**PS**: Please suggest a better way (if you have) to solve this issue. I would be more than happy to implement it. :)

HardikModha commented:
@bfred-it @sindresorhus Can you please take a look? Thanks!

bfred-it commented on source/content.ts:
```diff
@@ -113,6 +114,7 @@ import './features/hide-tips.css';
 import './features/hide-readme-header.css';
 import './features/hide-obvious-tooltips.css';
 import './features/clean-discussions.css';
+import './features/what-changed-since-previous-release.css';
```
This should be inside the feature itself

HardikModha replied:
Is it needed? The styles are getting applied irrespective of it. Also, none of the current features has these CSS files imported.

bfred-it replied:
Yes it’s needed, the JS features should “load” their own CSS. Those in content.ts are CSS-only features

Look better, there are quite a few feature that load their own CSS

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,175 @@
+	// To extract the commit hash from links like "/facebook/react/commit/92a1d8feac32d03ab5ea6ac13ae4941f6ae93b54"
+	const commitRegExp = new RegExp(`/${ownerName}/${repoName}/commit/([0-9a-f]{5,40})`);
+
+	const allTagsAnchor = [...select.all<HTMLAnchorElement>(getTagSelector())];
```
`select.all` already returns an array

HardikModha replied:
🤦‍♂ There was some Typescript error and it wasn't going away but now it's gone. 💭 Removed the use of spread.

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,175 @@
+	const tagRegExp = new RegExp(`/${ownerName}/${repoName}/releases/tag/(.*)`);
+
+	// To extract the commit hash from links like "/facebook/react/commit/92a1d8feac32d03ab5ea6ac13ae4941f6ae93b54"
+	const commitRegExp = new RegExp(`/${ownerName}/${repoName}/commit/([0-9a-f]{5,40})`);
```
These links never point to a different repo, so you just need

```suggestion
	const tagRegExp = /releases\/tag\/(.*)$/;
	const commitRegExp = /commit\/([0-9a-f]{40})$/;
```

HardikModha commented:
> * `getCommitIdSelector` and `getTagSelector` can be skipped if you just use `commitAnchorSelector` and `tagAnchorSelector` respectively? This lets you skip a lot of code and even `isReleasesListPage` and `isTagsListPage` themselves

Done. Simplified the commit id and tag selectors. Also, removed the checks for `isReleaseListPage` and `isTagsListPage` by using the common classname as selector which was present on both the pages.

> * the code after `allTagsAnchor` and `nextPageAllTagsAnchor` is nearly identical and can probably be merged.

Now, Appending the next page tags into the same array and iterating only once. Check for the last tag is also removed.

> Also always use `for-of` and not a regular `for`

Here, I need an index of the item so that I can start looking for the tags to match just after that index. You can take a look from Line no 36 and Line no 62. 

> Position: I was expecting this link to be on the `commit | zip | tar.gz` line and something more clear than `...` (I didn't even see the link at first, that `...` is generally used to expand the content)

Initially, I had chosen the same. The same line as commit | zip | tar.gz and for the releases just below the commit hash. (As displayed in the screenshot)

![1ec2a437-cc8e-45eb-97f7-4d15ac433700](https://user-images.githubusercontent.com/22439276/56993431-1318eb00-6bba-11e9-930d-f336586359d0.png)

But there was some miscommunication due to https://github.com/sindresorhus/refined-github/issues/966#issuecomment-487345368 and I put the link on `...` 😓.

HardikModha commented:
@bfred-it Please take a look at new changes and suggest me if anything still needs to be changed. :)

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,127 @@
+		const previousTagIndex = getPreviousTagIndex(index + 1, allCommitIds[index], allCommitIds, allTags[index], allTags);
+
+		if (previousTagIndex !== -1) {
+			allCommitIdsAnchor[index].parentElement!.parentElement!.append(getTagComparisonLink(allTags[previousTagIndex], allTags[index]));
```
Use `.closest` instead of multiple `parentElement`

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,127 @@
+	const [, tagTwoNameSpace = ''] = tag2.match(namespaceRegex) || [];
+
+	return tagOneNameSpace === tagTwoNameSpace;
+};
```
If I’m understanding this correctly, hierarchy and namespace are actually part of the same thing. What we’re looking for here is SOMETHING@version. It does not matter what the shape of SOMETHING is, you just need to make sure that SOMETHING is the same for both links, if it exists. Example:

```js
function doesNamespaceMatch() {
  const [namespace1] = tag1.split(/@[^@]+$|^/); // I think this regex will always keep the version at index 1 of the array
  const [namespace2] = tag2.split(/@[^@]+$|^/);
  return namespace1 === namespace2;
}
```

HardikModha replied:
Much better. 🎉

bfred-it replied:
I'm working on the regex, it doesn't what I had hoped

bfred-it replied:
Ok, now it sort of does. It relies on the assumption that there's no such thing as `v1.2@v1.2` tag, so even if in a namespace-less tag the version is at the index `0`, it won't affect the comparison:

<img width="311" alt="" src="https://user-images.githubusercontent.com/1402241/57007585-50ae6000-6c1c-11e9-87a7-a49dcccb3e70.png">

HardikModha replied:
Hey @bfred-it, How did you know I was doing the same? 🤔 Simplifying the regex. I just changed regex from `/@[^@]+$|^/` to `/@[^@]+$/`. You spyin' on me? 😂

bfred-it replied:
I had added `|^` thinking that it would split `v1.2` into `['', 'v1.2`]`, but that only broke the `@user/module@version`:

<img width="313" alt="Screenshot 2019-05-01 at 14 26 35" src="https://user-images.githubusercontent.com/1402241/57007697-34f78980-6c1d-11e9-888e-c637c79e16bf.png">

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	for (let index = 0; index < allCommitIdsAnchor.length; index++) {
+		const previousTagIndex = getPreviousTagIndex(index + 1, allCommitIds[index], allCommitIds, allTags[index], allTags);
```
You’re passing the same information twice. 
```suggestion
		const previousTagIndex = getPreviousTagIndex(index + 1, allCommitIds, allTags);
```

You’re already passing `index` so create the missing variables in the function. This avoid having a 5-parameter function (too many parameters)

bfred-it commented on source/features/what-changed-since-previous-release.css:
```diff
@@ -0,0 +1,14 @@
+}
+
+.rgh-what-changed {
+	cursor: pointer;
```
```suggestion
```

It's already a link

HardikModha replied:
CSS file removed.

bfred-it commented on source/features/what-changed-since-previous-release.css:
```diff
@@ -0,0 +1,14 @@
+}
+
+.release > div:first-child > ul > li.rgh-what-changed {
+	margin-bottom: 4px;
```
See if you can replace these styles with GitHub's own margin classes, like `mb-1` and `d-inline-block`. Maybe you can apply them without any conditions (i.e. just use `className="mb-1 mr-1 mt-1"` on all)

HardikModha replied:
Much better. Everything just reduced to `d-inline-block mb-1 mt-1 f6`. 
Got rid of the `css` file.

bfred-it replied:
Don't forget to push your changes, I still see the file :)

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,103 @@
+	return prevTagIndex;
+};
+
+const doesNamespaceMatch = (tag1: string, tag2: string): boolean => tag1.split(/@[^@]+$/)[0] === tag2.split(/@[^@]+$/)[0];
```
```suggestion
const doesNamespaceMatch = (tag1: string, tag2: string): boolean =>
	tag1.split(/@[^@]+$/)[0] === tag2.split(/@[^@]+$/)[0];
```

HardikModha replied:
Can we add a lint rule to avoid this in the future? https://eslint.org/docs/rules/max-len

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	return prevTagIndex;
+};
```
I think this defaults to the last tag on the page (or the next page), it never returns `-1`

If it cannot find a matching namespace, it should just use the next non-same-commit tag.

HardikModha replied:
It can return `-1` when all the tags on the page are having the same commit id.

```
if (allCommitIds[i] === allCommitIds[index]) {
	continue;
}
```

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,103 @@
+		const previousTagIndex = getPreviousTagIndex(index, allCommitIds, allTags);
+
+		if (previousTagIndex !== -1) {
+			allCommitIdsAnchor[index].closest('ul')!.append(getTagComparisonLink(allTags[previousTagIndex], allTags[index]));
```
You can inline `getTagComparisonLink` here (i.e. just create JSX inside `.append()`)

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,111 @@
+	return (
+		<li className="rgh-what-changed">
+			<a className="muted-link text-mono" href={`/${ownerName}/${repoName}/compare/${prevTag}...${nextTag}`}>
+				<div className="d-inline-block mr-1">{diff()}</div>compare
```
```suggestion
				{diff()} Changelog
```

Other links on that line just rely on the space between the icon and the text

bfred-it commented on source/features/what-changed-since-previous-release.tsx:
```diff
@@ -0,0 +1,111 @@
+};
+
+features.add({
+	id: 'what-changed-since-previous-release',
```
```suggestion
	id: 'tag-changelog-link',
```

I think this is a clear name for this feature. Also the file should be renamed

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,111 @@
+
+	return (
+		<li className="rgh-tag-changelog-link">
+			<a className="muted-link text-mono" href={`/${ownerName}/${repoName}/compare/${prevTag}...${nextTag}`}>
```
```suggestion
			<a className="muted-link text-mono tooltipped tooltipped-n" aria-label={'See changes since ' + prevTag} href={`/${ownerName}/${repoName}/compare/${prevTag}...${nextTag}`}>
```

HardikModha replied:
Love it, man! Thanks! Great suggestion!

bfred-it commented:
The current version is buggy, it's skipping version 29 (look at the status bar):

<img width="453" alt="" src="https://user-images.githubusercontent.com/1402241/57009119-4cd40b00-6c27-11e9-9bff-87512ac41b49.png">

https://github.com/sindresorhus/refined-github/releases

HardikModha commented:
> The current version is buggy, it's skipping version 29 (look at the status bar):
> 
> <img alt="" width="453" src="https://user-images.githubusercontent.com/1402241/57009119-4cd40b00-6c27-11e9-9bff-87512ac41b49.png">
> 
> https://github.com/sindresorhus/refined-github/releases

Technically it isn't. The tag is on the same commit id so it looks for the next tag on different commit id.

bfred-it commented:
> Technically it isn't.

D'oh. True. #1997 coming soon

bfred-it commented:
I'm not seeing the link on https://github.com/sindresorhus/refined-github/releases/tag/19.4.30.1910

HardikModha commented:
@bfred-it Seems like you missed screenshots from https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-488116969.

For releases, the link is displayed just below the commit id on the left side.

![changelog-link](https://user-images.githubusercontent.com/22439276/57009416-1640c500-6c14-11e9-9059-b689d6cd091b.png)

bfred-it commented:
Seems like it's not working

<img width="758" alt="Screenshot 2019-05-01 at 15 53 33" src="https://user-images.githubusercontent.com/1402241/57009456-49418380-6c29-11e9-855d-1fdafac8b24e.png">

HardikModha commented:
@bfred-it Just confirming, You are on the latest changes. Right? Also, Can you check the console if there is anything? It seems to work fine on my end. Tried for multiple tags.

bfred-it commented:
Ah yeah, this is due to Firefox requiring `location.origin` on `fetch`

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,112 @@
+
+	const {ownerName, repoName} = getOwnerAndRepo();
+
+	return fetchDom(`/${ownerName}/${repoName}/releases?after=${tag}`);
```
```suggestion
	return fetchDom(`${location.origin}/${ownerName}/${repoName}/releases?after=${tag}`);
```

HardikModha commented:
> I'm not seeing the link on [`19.4.30.1910` (release)](https://github.com/sindresorhus/refined-github/releases/tag/19.4.30.1910) and [bfred-it/insert-text-textarea@`v1.0.3` (release)](https://github.com/bfred-it/insert-text-textarea/releases/tag/v1.0.3)

It's also working on https://github.com/bfred-it/insert-text-textarea/releases/tag/v1.0.3. Are you on Chrome right?

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,112 @@
+
+const extractValuesFromPathname = (anchors: HTMLAnchorElement[], regexp: RegExp): string[] => {
+	return anchors.map((anchor: HTMLAnchorElement): string => {
+		const [, value] = anchor.pathname.match(regexp)!;
```
```suggestion
		return anchor.textContent!.trim();
```

You might have to adjust the selector to select these links:

<img width="127" alt="" src="https://user-images.githubusercontent.com/1402241/57009786-5f504380-6c2b-11e9-8c9b-13e9518a53f2.png">

bfred-it commented:
You can use this:

<img width="511" alt="" src="https://user-images.githubusercontent.com/1402241/57081611-ad4a7180-6d27-11e9-9cb6-c54ec1ac18bb.png">

HardikModha commented:
Seems like the changes you made has introduced a bug.

![Screen Shot 2019-05-02 at 11 58 20 PM](https://user-images.githubusercontent.com/22439276/57097780-3f597680-6d36-11e9-8275-75652f94988b.png)

It should compare with `@parcel/integration-tests@1.12.2` not `@parcel/integration-tests@1.12.1`.

HardikModha commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,93 @@
+	// https://github.com/facebook/react/releases (releases list)
+	// https://github.com/parcel-bundler/parcel/releases (releases list without release notes)
+	const tagElements = select.all('.f1 [href*="/releases/tag"], .commit-title [href*="/releases/tag"]', documents);
+	const commitElements = select.all('.muted-link[href*="/commit/"]', documents);
```
This selector is not reliable. It is selecting multiple links on the single tag release page causing the wrong tag to be displayed as the previous tag. We'll need to change it.

HardikModha replied:
The same issue is also occurring on the release page where the tag is an actual release with the release notes. 

Apart from that, The feature has broken on such releases page due to index out of bounds. 

![Screen Shot 2019-05-03 at 12 18 45 AM](https://user-images.githubusercontent.com/22439276/57099248-9876d980-6d39-11e9-86a9-d5dcbe9e0998.png)

sindresorhus commented:
Can you add this feature file to https://github.com/sindresorhus/refined-github/blob/master/.github/CODEOWNERS ?

sindresorhus commented:
It doesn't work for releases with a custom title. It should not use the release title, but rather the Git tag connected to the release. Example: https://github.com/avajs/ava/releases/tag/v2.0.0-beta.1 (Click "Changelog").

sindresorhus commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,93 @@
+/*
+This feature adds a compare link on each releases/tags/single tag page so that you can see
+what has changed since the previous release. If the tags are namespaced then it tries to
+get the previous release of the same namespaced tag.
```
Don't hard-wrap.

sindresorhus commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,93 @@
+		}
+
+		// Ensure that they have the same namespace. e.g. `parcel@1.2.4` and `parcel@1.2.3`
+		if (tags[i].split(/@[^@]+$/)[0] === tags[index].split(/@[^@]+$/)[0]) {
```
The regex should be moved to a variable with a descriptive name and reused for both of these.

bfred-it commented:
> It doesn't work for releases with a custom title. It should not use the release title, but rather the Git tag connected to the release. Example: [avajs/ava@`v2.0.0-beta.1` (release)](https://github.com/avajs/ava/releases/tag/v2.0.0-beta.1) (Click "Changelog").

Indeed, there are so many variations to the dom I ended up forgetting about this one.

One more thing missing is a link in the second meta box (it's duplicate in the dom for responsiveness). For this reason (and to avoid this awkward "double-select that has to match the same number of items"), I'd suggest another way to do this, something like:


```js
const tags = select.all('each tag’s main element').map(element => ({
	element,
	tag: element.querySelector('...'),
	commit: element.querySelector('...')
}));
```

This way there will be **one** array of tags/releases with content already parsed. Later, when adding the buttons, you can run something like:

```js
if (previousTag !== false) {
	const containers = tag.element.querySelectorAll('all the places that need the link');
	for (const container of containers) {
		container.append(
			...JSX...
		);
	}
}
```

HardikModha commented:
> For this reason (and to avoid this awkward "double-select that has to match the same number of items"), I'd suggest another way to do this, something like:

We are already avoiding the duplicate link right now by reverting back to the older selector. https://github.com/sindresorhus/refined-github/pull/1998/commits/67822230c90af73e2d4459d680028d53b82e4d05
Other than that, Do we also need to show the "Changelog" link for the smaller screens?

HardikModha commented:
Also, As Sindre pointed out a case, we'll need to revert back to the old regex-based approach to extract tag. 
https://github.com/sindresorhus/refined-github/blob/e38619ab55228937b8a5c42d21c965862eac68e7/source/features/what-changed-since-previous-release.tsx#L15

bfred-it commented:
> Do we also need to show the "Changelog" link for the smaller screens?

It would be ideal, also because it's not "smaller screens" but "smaller windows", e.g. side-by-side

> Also, As Sindre pointed out a case, we'll need to revert back to the old regex-based approach to extract tag.

Why? The tag is available in the dom, we just need to update the way it parses the dom as suggested. 

As we've seen, any changes to the selector could cause subtle errors that make it mismatch commits and tags, so fetching by "tag container" first would be preferred in every case.

HardikModha commented:
I just thought extracting from the link would be much safer. The tag name can be customized so.

e.g. If on the parcel release page they use the custom tag name, we don't have another way to extract actual tag name from the DOM. We'll have to use the regex to extract the name from the link. 

![Screen Shot 2019-05-03 at 11 40 33 PM](https://user-images.githubusercontent.com/22439276/57156972-0df7ae00-6dfd-11e9-8052-2c9e68f91d7c.png)

For the releases with release notes, we can extract it from the container present on the left side. But then why follow two separate approaches? We can just stick to the regex.

bfred-it commented:
Alright, regex it is then :)

https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-488970348 still applies though

HardikModha commented:
You mean displaying it on the smaller window or the approach that you suggested? or both? 😅

HardikModha commented:
Completely unrelated to the current topic, https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-489183075 which I just made earlier (In the current thread), I cannot see the content when the scrollbar is present. There should be some minimum height enforced. What do you think?

bfred-it commented:
> You mean displaying it on the smaller window or the approach that you suggested? or both? 😅

The approach I suggested will make it appear on the smaller window

> Completely unrelated to the current topic, [#1998 (comment)](https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-489183075) which I just made earlier (In the current thread), I cannot see the content when the scrollbar is present. There should be some minimum height enforced. What do you think?

Already a fix by @notlmn coming in #1989 😍

HardikModha commented:
So I followed the kind of same approach that @bfred-it suggested in (https://github.com/sindresorhus/refined-github/pull/1998/#issuecomment-488970348) with a little difference. I'm not maintaining an object of `element`, `tag` and `commit` instead it's just an array of subparts of the document which contains these tags. Then I'm iterating over each subpart to append the changelog link at all the places.

bfred-it commented:
> I'm not maintaining an object of `element`, `tag` and `commit`

Then we still have the same problem of 3 possibly-mismatching selectors. Please use the suggested solution.

> As we've seen, any changes to the selector could cause subtle errors that make it mismatch commits and tags, so fetching by "tag container" first would be preferred in every case.

HardikModha commented:
I'm already fetching the tag container first. All the searches for tag/commit hash are being performed on these containers. Can you please explain in which cases it will fail?

bfred-it commented:
It doesn't fail _now_ but it can **subtly** fail in the future like it did in https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-488781643

I just don't like reusing the same index on now-3 separate arrays generated with 3 long selectors and then stitched together. In my suggested solution there's one source of truth that we can easily inspect in case it fails

HardikModha commented:
Please correct me If I am wrong, the issue https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-488781643 was occurring because of the selector being too obvious (just `.muted-link[href*="/commit/"])` which resulted in false matches. That's why I have kept the selectors little more specific to avoid such false matches. 

Now regarding your last comment, I'm failing to understand how the solution I've implemented is different from what you are suggesting in the https://github.com/sindresorhus/refined-github/pull/1998#issuecomment-488970348. 
It, in fact, is doing what you are suggesting, just little differently. Again, correct me if I am wrong. 

You are suggesting to first select "all tags main elements" and then iterate over them to extract the tag and commit. I'm doing the same thing.

In my implementation, "each tag's main element" is `.release`, `.release-main-section:not(.commit)`, and `.Box-row .commit`. These three are the unique selectors which select the main container. 

1. `.release` => Releases with releases notes
2. `.release-main-section:not(.commit)` => Releases without release notes
3. `.Box-row .commit` => Tag on the tags listing. The `.Box-row` is important because the same `.commit` selector also exists as child in 1 & 2.

```
// Selecting all the main tag elements
const tagContainers = select.all('.release, .Box-row .commit, .release-main-section:not(.commit)', documents);
```

Now we have three different containers so in these containers tag and commit hash exists at different locations. So we need different selectors to extract the tag and commit hash. 

But as we know "releases without release notes" and "tags on listing page" have the same structure. We can use the same selector for them.

**Selectors for tag:**

1. `'h4.commit-title > a[href*="/releases/tag/"]` => Extracts the tag name from the title on tags listing and releases without release notes. 

2. `div.release-header .f1 > a[href*="/releases/tag/"]` => Extracts the tag name from the title from releases with release notes.

We just cannot use `a[href*="/releases/tag"]` as it _might_ return false matches.

**Selector for commit hash**

1. `.commit ul a[href*="/commit/"].muted-link` => In all three containers commit hash can be exracted using this single selector. 

```
// Extracting tag names from all (Passing tagContainers)
const tagElements = select.all<HTMLAnchorElement>('h4.commit-title > a[href*="/releases/tag/"], div.release-header .f1 > a[href*="/releases/tag/"]', tagContainers);

// Extracting commit hash from all (Passing tagContainers)
const commitElements = select.all<HTMLAnchorElement>('.commit ul a[href*="/commit/"].muted-link', tagContainers);
```

https://github.com/bfred-it/select-dom/blob/master/index.ts#L81 (Applies the passed selector in each individual container). 

Which is functionally same as, 

```
select.all('each tag’s main element').map(element => ({
	element,
	// These selectors can probably be very simple since they're already in the tag
	// and because they only select the first occurrence of the tag/commit link
	tag: element.querySelector('...'),
	commit: element.querySelector('...')
}));
```

The only difference I can see is that in my implementation we are doing the query for "tag" and "commit" separately.

And now at the last, we select all the places where we need the "Changelog link".

```
select.all('.commit > ul.f6, .commit > .release-header > ul, div:first-child > ul', tagContainer);
```

Don't get me wrong. I can easily convert my implementation to follow the _exact_ steps that you are suggesting but I just want to understand what am I missing here?

HardikModha commented:
Using your suggested approach as for edge case  (https://github.com/sindresorhus/refined-github/pull/1998/commits/5c0d57f11aabffde2c4b611dc4205538f5cfba18) my approach silently failed but your approach threw an error which I really liked. Also in your suggested approach, we only need to iterate only once to extract tag and commit details. (Removed the big comment which now I think is unnecessary.)

Let me know if any more changes are needed. Thanks :)

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,114 @@
+		}
+
+		// Ensure that they have the same namespace. e.g. `parcel@1.2.4` and `parcel@1.2.3`
+		if (allTags[i].tag.split(namespaceRegex)[0] === allTags[index].tag.split(namespaceRegex)[0]) {
```
Since `namespaceRegex` is a separate variable again, we might as well turn it into a global function, e.g.

```js
const getNameSpace = tag => tag.split(/@[^@]+$/)[0];
```

```suggestion
		if (getNameSpace(allTags[i].tag) === getNameSpace(allTags[index].tag)) {
```

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,114 @@
+
+	const allTags: TagDetails[] = select.all(tagContainerSelector, documents).map(element => ({
+		element,
+		tag: decodeURIComponent(select<HTMLAnchorElement>(tagSelector, element)!.pathname.match(tagRegex)![1]),
```
I think this isn't needed, also because the tag will end up back into a URL anyway
```suggestion
		tag: select<HTMLAnchorElement>(tagSelector, element)!.pathname.match(tagRegex)![1],
```

HardikModha replied:
We are also showing the tooltip so if we don't `decodeURIComponent` it will display tooltip text in the encoded form.

bfred-it replied:
Indeed. Then maybe we need to move `decodeURIComponent` to the tooltip because we need the tag to be urlencoded in the URL

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,114 @@
+	const allTags: TagDetails[] = select.all(tagContainerSelector, documents).map(element => ({
+		element,
+		tag: decodeURIComponent(select<HTMLAnchorElement>(tagSelector, element)!.pathname.match(tagRegex)![1]),
+		commit: select('.commit ul a[href*="/commit/"].muted-link', element)!.textContent!.trim()
```
Can this be simplified now that it's limited to the `tagContainer`? `select` will only pick the only link anyway, we don't have a problem like before.

```suggestion
		commit: select('[href*="/commit/"]', element)!.textContent!.trim()
```

HardikModha replied:
Same as https://github.com/sindresorhus/refined-github/pull/1998#discussion_r281033111

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,114 @@
+	// https://github.com/facebook/react/tags (tags list)
+	// https://github.com/facebook/react/releases (releases list)
+	// https://github.com/parcel-bundler/parcel/releases (releases list without release notes)
+	const tagSelector = 'h4.commit-title > a[href*="/releases/tag/"], div.release-header .f1 > a[href*="/releases/tag/"]';
```
Can this be simplified now that it's limited to the `tagContainer`? `select` will only pick the only link anyway, we don't have a problem like before.
```suggestion
	const tagSelector = '[href*="/releases/tag/"]';
```

HardikModha replied:
Yes, you are right the container is limited to `tagContainer` but this `tagContainer` can be either "release with notes" or "release without notes". So with the release with notes false matches can occur in the description.

bfred-it replied:
The description comes after the header, and the `/releases/tag/` link is in the header, so it's matched first

bfred-it replied:
Yeah, this is what I was referring to. One long `tagContainer` selector and the rest can be more simple

HardikModha commented:
> Thank you and sorry for all the reviews

Haha...No need for sorry. I learned a lot through your and Sindre's hardcore reviews. Most of the time I (or any new person) is focused on making something to work and we forget some scenarios or make some silly mistakes. That's where you guys come into the picture. Just making something to work is not enough. I really appreciate your and Sindre's attitude towards perfectness. :)

bfred-it commented on source/features/tag-changelog-link.tsx:
```diff
@@ -0,0 +1,113 @@
+import select from 'select-dom';
+import features from '../libs/features';
+import fetchDom from '../libs/fetch-dom';
+import {diff} from '../libs/icons';
```
```suggestion
import * as diff from '../libs/icons';
```

I think it's clearer to use this as `icons.diff()` later later than `diff()`

bfred-it commented on source/features/tag-changelog-link.css:
```diff
@@ -0,0 +1,3 @@
+.release div.release-header > ul > li {
+	flex-grow: 0 !important;
```
I understand that this is to avoid this issue:

<img width="664" alt="" src="https://user-images.githubusercontent.com/1402241/57436361-183d0080-7272-11e9-98bc-1d6b7f658478.png">


But this isn't right:

<img width="689" alt="Screenshot 2019-05-09 at 15 50 10" src="https://user-images.githubusercontent.com/1402241/57436439-3f93cd80-7272-11e9-8ea5-5d65802a5643.png">


The solution might be a bit ugly, but I think you should:

1. add the link after that `flex-auto`
2. remove the `flex-auto` class
3. add the `flex-auto` class to the `Changelog` link

Maybe you can do this like:

```js
if (position.is('.flex-auto')) {
  position.classList.remove('flex-auto');
  button.classList.add('flex-auto');
  position.after(button)
}
```

HardikModha commented:
> Also, perhaps this should use `tiny-version-compare` to ensure that the tag we're comparing to is actually lower than the current one.
> 
> For example on: https://github.com/tensorflow/tensorflow/releases
> The tag right after `v1.12.1` is `TensorFlow 2.0.0-alpha0` and the generated link shouldn't be:
> 
> [tensorflow/tensorflow@v2.0.0-alpha0...v1.12.1](https://github.com/tensorflow/tensorflow/compare/v2.0.0-alpha0...v1.12.1)
> 
> but it should compare it against `v1.12.0`, further down the page
> 
> [tensorflow/tensorflow@v1.12.0...v1.12.1](https://github.com/tensorflow/tensorflow/compare/v1.12.0...v1.12.1)

Yep, you are right. I am aware of this limitation and have already mentioned it in a description. I didn't look for any packages (pure out of laziness maybe) 😅. Thanks for the `tiny-version-compare` suggestion. This wekend, I'm out so won't be able to work on it. Next week I'll give it a try.

bfred-it commented:
Hallelujah! Finally we merged it. Thanks for this feature!

HardikModha commented:
Woahhh, Thanks for refining the feature. ❤️ I just returned today and saw this merged!!! 🎉 

But seems like there is a small issue. Tags are not being compared with a namespace.  

![Screen Shot 2019-05-14 at 8 48 20 PM](https://user-images.githubusercontent.com/22439276/57709975-a60d5700-7689-11e9-8c23-a89d600798ea.png)
It should compare `parcel-bundler` with it's matching the previous tag, not with the `parcel-integration`. The tag name is not being decoded using before extracting `version` and `namespace` using `parseTags`.

HardikModha commented:
Also, I noticed a behavior with `version` aware comparison. It is comparing the first previous tag and not the closest matching previous release. 

e.g. On https://github.com/flutter/flutter/releases?after=v1.5.1 releases page. 

![Screen Shot 2019-05-14 at 8 59 04 PM](https://user-images.githubusercontent.com/22439276/57710965-86772e00-768b-11e9-9620-7b2a52dba187.png) 

It should compare `v1.4.17` with `v1.4.16` and not `v1.4.9-hotfix.1`. What do you think?
